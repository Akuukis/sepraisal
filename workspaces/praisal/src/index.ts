export * from './Praisal'
export * from './PraisalManager'
export * from './models'
export * from './unzipCachedSbc'
