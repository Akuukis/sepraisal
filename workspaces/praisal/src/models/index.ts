
export * from './Cube.ts'
export * from './Blueprint.ts'
export * from './Component.ts'
export * from './Ingot.ts'
export * from './Ore.ts'
export * from './Orientation.ts'


export * from './Group.ts'
