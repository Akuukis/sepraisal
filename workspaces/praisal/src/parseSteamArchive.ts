import { Stream } from 'stream'
import * as unzip from 'unzip'


export const parseSteamArchive = async (readStream: Stream): Promise<string> => {
    let result: string

    return (new Promise<string>((resolve, reject) => {
        readStream
            .on('error', reject)
            .pipe(unzip.Parse())
                .on('entry', (entry: unzip.Entry) => {
                    try {
                        const fileName = entry.path
                        const chunks: Buffer[] = []

                        entry.on('data', (chunk: Buffer) => {
                            chunks.push(chunk)
                        })

                        entry.on('end', () => {
                            if(fileName.endsWith('.sbc')) {
                                result = Buffer.concat(chunks).toString()
                            }
                        })
                    } catch(err) {
                        reject(err)
                    }
                })
                .on('error', reject)
                .on('close', () => {
                    resolve(result)
                })
    }))

}
