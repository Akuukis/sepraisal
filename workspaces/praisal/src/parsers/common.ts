/**
 * Not the best config, but as close to xml2js as possible.
 */
export const PARSE_CONFIG = {
    attributeNamePrefix: '',
    attrNodeName: '$',
    ignoreAttributes: false,
    arrayMode: (tagName: string): boolean => tagName !== '$',
}
