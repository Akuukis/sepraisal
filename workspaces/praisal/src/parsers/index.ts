export * from './parseBlueprint'
export * from './parseBlueprintSbc'
export * from './parseComponentSbc'
export * from './parsePhysicalItemsSbc'
