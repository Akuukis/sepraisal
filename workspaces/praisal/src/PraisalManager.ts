import { BLOCK_GROUPS, ObservableMap, VENDOR_MOD } from '@sepraisal/common'
import { action, runInAction } from 'mobx'

import { Blueprint, Component, Cube, Group, Ingot, Ore } from './models'
import {
    IParseBlueprintSbc,
    IParseComponentSbc,
    IParseCubeBlocksSbc,
    IParsePhysicalItemsSbc,
    parseBlueprintSbc,
    parseComponentSbc,
    parseCubeBlocksSbc,
    parsePhysicalItemsSbc,
} from './parsers'
import { Praisal } from './Praisal'

// tslint:disable-next-line: min-class-cohesion
export class PraisalManager {
    private readonly blueprintSbcs = new ObservableMap<IParseBlueprintSbc>()
    private readonly physicalItemSbcs = new ObservableMap<IParsePhysicalItemsSbc>()
    private readonly componentSbcs = new ObservableMap<IParseComponentSbc>()
    private readonly cubeBlocksSbcs = new ObservableMap<IParseCubeBlocksSbc>()

    public readonly components = new ObservableMap<Component>()
    public readonly ingots = new ObservableMap<Ingot>()
    public readonly ores = new ObservableMap<Ore>()

    public readonly cubes = new ObservableMap<Cube>()

    public readonly groups = new ObservableMap<Group>()

    public async addBlueprintsSbc(blueprintsSbc: string, mod: VENDOR_MOD): Promise<void> {
        const blueprintSbcs = await parseBlueprintSbc(blueprintsSbc, mod)
        runInAction('PraisalManager.addBlueprintsSbc', () => {
            this.blueprintSbcs.merge(blueprintSbcs.map((sbc) => [sbc.fullType, sbc]))
        })
    }
    public async addComponentsSbc(componentsSbc: string, mod: VENDOR_MOD): Promise<void> {
        const componentsSbcs = await parseComponentSbc(componentsSbc, mod)
        runInAction('PraisalManager.addComponentsSbc', () => {
            this.componentSbcs.merge(componentsSbcs.map((sbc) => [sbc.fullType, sbc]))
        })
    }
    public async addPhysicalItemsSbc(physicalItemsSbc: string, mod: VENDOR_MOD): Promise<void> {
        const physicalItemsSbcs = await parsePhysicalItemsSbc(physicalItemsSbc, mod)
        runInAction('PraisalManager.addPhysicalItemsSbc', () => {
            this.physicalItemSbcs.merge(physicalItemsSbcs.map((sbc) => [sbc.fullType, sbc]))
        })
    }
    public async addCubeBlocksSbc(cubeBlocksSbc: string, mod: VENDOR_MOD): Promise<void> {
        const cubeBlocksSbcs = await parseCubeBlocksSbc(cubeBlocksSbc, mod)
        runInAction('PraisalManager.addCubeBlocksSbc', () => {
            this.cubeBlocksSbcs.merge(cubeBlocksSbcs.map((sbc) => [`${String(sbc.Id[0].TypeId[0])}/${sbc.Id[0].SubtypeId[0]}`, sbc]))
        })
    }

    @action public build(): void {
        this.buildOres()
        this.buildIngots()
        this.buildComponents()
        this.buildCubes()
    }

    @action private buildOres() {
        const ore = Ore.fromSbcs(this.physicalItemSbcs)
        this.ores.merge(ore.map((ore) => [ore.title, ore]))
    }

    @action private buildIngots() {
        const ingots = Ingot.fromSbcs(this.physicalItemSbcs, this.blueprintSbcs)
        this.ingots.merge(ingots.map((ingot) => [ingot.title, ingot]))
    }

    @action private buildComponents() {
        const components = Component.fromSbcs(this.blueprintSbcs, this.componentSbcs)
        this.components.merge(components.map((component) => [component.title, component]))
    }

    @action private buildCubes() {
        const cubes = Cube.fromSbcs(this.components, this.cubeBlocksSbcs)
        this.cubes.merge(cubes.map((component) => [component.title, component]))
    }

    @action public addGroups(groups2: typeof BLOCK_GROUPS): void {
        groups2.forEach((groupDto) => this.groups.set(groupDto.title, new Group(groupDto)))
    }

    @action public praise(blueprint: Blueprint): Praisal {
        const praisal = new Praisal(blueprint, this, this.groups)

        return praisal
    }

    public async praiseSbc(sbc: string): Promise<Praisal> {
        const blueprint = await Blueprint.parseSbc(sbc, this.cubes)

        return this.praise(blueprint)
    }
}
