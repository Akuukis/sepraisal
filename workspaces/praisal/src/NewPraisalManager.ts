import { BLOCK_GROUPS, VENDOR_MOD } from '@sepraisal/common'
import { readFileSync } from 'fs'
import { join } from 'path'

import { PraisalManager } from './PraisalManager'

export const VENDOR_DIR = join(__dirname, '..', 'vendor')

export const NewPraisalManager = (): () => Promise<PraisalManager> => {
    const cubeBlocksSbcs = [
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Armor.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Automation.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Communications.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Control.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Doors.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Energy.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Extras.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Gravity.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Interiors.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_LCDPanels.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Lights.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Logistics.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Mechanical.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Medical.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Production.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Symbols.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Thrusters.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Tools.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Utility.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Weapons.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Wheels.sbc')],
        [VENDOR_MOD.VANILLA     , join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'CubeBlocks', 'CubeBlocks_Windows.sbc')],

        [VENDOR_MOD.DECORATIVE_1        , join(VENDOR_DIR, VENDOR_MOD.DECORATIVE_1      , 'CubeBlocks.sbc')],
        [VENDOR_MOD.DECORATIVE_2        , join(VENDOR_DIR, VENDOR_MOD.DECORATIVE_2      , 'CubeBlocks.sbc')],
        [VENDOR_MOD.ECONOMY             , join(VENDOR_DIR, VENDOR_MOD.ECONOMY           , 'CubeBlocks.sbc')],
        [VENDOR_MOD.FROSTBITE           , join(VENDOR_DIR, VENDOR_MOD.FROSTBITE         , 'CubeBlocks.sbc')],
        [VENDOR_MOD.SPARKSOFTHEFUTURE   , join(VENDOR_DIR, VENDOR_MOD.SPARKSOFTHEFUTURE , 'CubeBlocks.sbc')],
    ].map(([mod, path]) => [mod, readFileSync(path).toString()] as [VENDOR_MOD, string])
    const componentsSbc = readFileSync(join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'Components.sbc')).toString()
    const blueprintsSbc = readFileSync(join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'Blueprints.sbc')).toString()
    const physicalItemsSbc = readFileSync(join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'PhysicalItems.sbc')).toString()

    return async () => {
        const sepraisal = new PraisalManager()
        await sepraisal.addPhysicalItemsSbc(physicalItemsSbc, VENDOR_MOD.VANILLA)
        await sepraisal.addBlueprintsSbc(blueprintsSbc, VENDOR_MOD.VANILLA)
        await sepraisal.addComponentsSbc(componentsSbc, VENDOR_MOD.VANILLA)
        for(const [mod, cubeBlocksSbc] of cubeBlocksSbcs) await sepraisal.addCubeBlocksSbc(cubeBlocksSbc, mod)
        sepraisal.build()
        sepraisal.addGroups(BLOCK_GROUPS)

        return sepraisal
    }
}
