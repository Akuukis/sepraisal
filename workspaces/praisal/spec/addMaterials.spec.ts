import { VENDOR_MOD } from '@sepraisal/common/src'
import { readFileSync } from 'fs'
import { join } from 'path'

import { PraisalManager } from '../src'


// tslint:disable: no-duplicate-string
const VENDOR_DIR = join(__dirname, '..', 'vendor')

const physicalItemsSbcVanilla = readFileSync(join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'PhysicalItems.sbc')).toString()
const materialsSbcVanilla = readFileSync(join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'Blueprints.sbc')).toString()
const componentsSbcVanilla = readFileSync(join(VENDOR_DIR, VENDOR_MOD.VANILLA, 'Components.sbc')).toString()
const physicalItemsSbcEconomy = readFileSync(join(VENDOR_DIR, VENDOR_MOD.ECONOMY, 'PhysicalItems.sbc')).toString()
const materialsSbcEconomy = readFileSync(join(VENDOR_DIR, VENDOR_MOD.ECONOMY, 'Blueprints.sbc')).toString()
const componentsSbcEconomy = readFileSync(join(VENDOR_DIR, VENDOR_MOD.ECONOMY, 'Components.sbc')).toString()

let sepraisal: PraisalManager

const testMaterials = () => {
    test('should build all ores', async () => {
        expect(sepraisal.ores.size).toMatchSnapshot()
        expect([...sepraisal.ores.keys()]).toMatchSnapshot()
        expect([...sepraisal.ores.values()]).toMatchSnapshot()
    })
    test('should build all ingots', async () => {
        expect(sepraisal.ingots.size).toMatchSnapshot()
        expect([...sepraisal.ingots.keys()]).toMatchSnapshot()
        expect([...sepraisal.ingots.values()]).toMatchSnapshot()
    })
    test('should build all components', async () => {
        expect(sepraisal.components.size).toMatchSnapshot()
        expect([...sepraisal.components.keys()]).toMatchSnapshot()
        expect([...sepraisal.components.values()]).toMatchSnapshot()
    })
}

describe('PraisalManager with vanilla materials only', () => {
    beforeEach(async () => {
        sepraisal = new PraisalManager()
        await sepraisal.addPhysicalItemsSbc(physicalItemsSbcVanilla, VENDOR_MOD.VANILLA)
        await sepraisal.addBlueprintsSbc(materialsSbcVanilla, VENDOR_MOD.VANILLA)
        await sepraisal.addComponentsSbc(componentsSbcVanilla, VENDOR_MOD.VANILLA)
        sepraisal.build()
    })
    testMaterials()
})
describe('PraisalManager with economy materials only', () => {
    beforeEach(async () => {
        sepraisal = new PraisalManager()
        await sepraisal.addPhysicalItemsSbc(physicalItemsSbcEconomy, VENDOR_MOD.ECONOMY)
        await sepraisal.addBlueprintsSbc(materialsSbcEconomy, VENDOR_MOD.ECONOMY)
        await sepraisal.addComponentsSbc(componentsSbcEconomy, VENDOR_MOD.ECONOMY)
        sepraisal.build()
    })
    testMaterials()
})
