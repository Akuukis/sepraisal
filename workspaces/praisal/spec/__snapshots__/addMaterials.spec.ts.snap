// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PraisalManager with economy materials only should build all components 1`] = `0`;

exports[`PraisalManager with economy materials only should build all components 2`] = `Array []`;

exports[`PraisalManager with economy materials only should build all components 3`] = `Array []`;

exports[`PraisalManager with economy materials only should build all ingots 1`] = `0`;

exports[`PraisalManager with economy materials only should build all ingots 2`] = `Array []`;

exports[`PraisalManager with economy materials only should build all ingots 3`] = `Array []`;

exports[`PraisalManager with economy materials only should build all ores 1`] = `0`;

exports[`PraisalManager with economy materials only should build all ores 2`] = `Array []`;

exports[`PraisalManager with economy materials only should build all ores 3`] = `Array []`;

exports[`PraisalManager with vanilla materials only should build all components 1`] = `22`;

exports[`PraisalManager with vanilla materials only should build all components 2`] = `
Array [
  "Component/Construction",
  "Component/Girder",
  "Component/MetalGrid",
  "Component/InteriorPlate",
  "Component/SteelPlate",
  "Component/SmallTube",
  "Component/LargeTube",
  "Component/Motor",
  "Component/Display",
  "Component/BulletproofGlass",
  "Component/Computer",
  "Component/Reactor",
  "Component/Thrust",
  "Component/GravityGenerator",
  "Component/Medical",
  "Component/RadioCommunication",
  "Component/Detector",
  "Component/Canvas",
  "Component/Explosives",
  "Component/SolarCell",
  "Component/PowerCell",
  "Component/Superconductor",
]
`;

exports[`PraisalManager with vanilla materials only should build all components 3`] = `
Array [
  Object {
    "health": 18,
    "mass": 8,
    "maxIntegrity": 30,
    "prerequisites": Object {
      "Ingot/Iron": 8,
    },
    "size": Object {
      "X": 0.2,
      "Y": 0.1,
      "Z": 0.1,
    },
    "subtype": "Construction",
    "time": 1,
    "type": "Component",
    "volume": 2,
  },
  Object {
    "health": 0,
    "mass": 6,
    "maxIntegrity": 15,
    "prerequisites": Object {
      "Ingot/Iron": 6,
    },
    "size": Object {
      "X": 0.05,
      "Y": 0.05,
      "Z": 1,
    },
    "subtype": "Girder",
    "time": 1,
    "type": "Component",
    "volume": 2,
  },
  Object {
    "health": 69,
    "mass": 6,
    "maxIntegrity": 30,
    "prerequisites": Object {
      "Ingot/Cobalt": 3,
      "Ingot/Iron": 12,
      "Ingot/Nickel": 5,
    },
    "size": Object {
      "X": 0.5,
      "Y": 0.5,
      "Z": 0.05,
    },
    "subtype": "MetalGrid",
    "time": 2,
    "type": "Component",
    "volume": 15,
  },
  Object {
    "health": 36,
    "mass": 3,
    "maxIntegrity": 15,
    "prerequisites": Object {
      "Ingot/Iron": 3,
    },
    "size": Object {
      "X": 0.5,
      "Y": 0.5,
      "Z": 0.05,
    },
    "subtype": "InteriorPlate",
    "time": 1,
    "type": "Component",
    "volume": 5,
  },
  Object {
    "health": 53,
    "mass": 20,
    "maxIntegrity": 100,
    "prerequisites": Object {
      "Ingot/Iron": 21,
    },
    "size": Object {
      "X": 0.5,
      "Y": 0.5,
      "Z": 0.01,
    },
    "subtype": "SteelPlate",
    "time": 1,
    "type": "Component",
    "volume": 3,
  },
  Object {
    "health": 18,
    "mass": 4,
    "maxIntegrity": 15,
    "prerequisites": Object {
      "Ingot/Iron": 5,
    },
    "size": Object {
      "X": 1,
      "Y": 0.04,
      "Z": 0.04,
    },
    "subtype": "SmallTube",
    "time": 1,
    "type": "Component",
    "volume": 2,
  },
  Object {
    "health": 22,
    "mass": 25,
    "maxIntegrity": 60,
    "prerequisites": Object {
      "Ingot/Iron": 30,
    },
    "size": Object {
      "X": 1,
      "Y": 0.2,
      "Z": 0.2,
    },
    "subtype": "LargeTube",
    "time": 2,
    "type": "Component",
    "volume": 38,
  },
  Object {
    "health": 14,
    "mass": 24,
    "maxIntegrity": 40,
    "prerequisites": Object {
      "Ingot/Iron": 20,
      "Ingot/Nickel": 5,
    },
    "size": Object {
      "X": 0.2,
      "Y": 0.2,
      "Z": 0.2,
    },
    "subtype": "Motor",
    "time": 3,
    "type": "Component",
    "volume": 8,
  },
  Object {
    "health": 12,
    "mass": 8,
    "maxIntegrity": 5,
    "prerequisites": Object {
      "Ingot/Iron": 1,
      "Ingot/Silicon": 5,
    },
    "size": Object {
      "X": 0.3,
      "Y": 0.4,
      "Z": 0.05,
    },
    "subtype": "Display",
    "time": 1,
    "type": "Component",
    "volume": 6,
  },
  Object {
    "health": 14,
    "mass": 15,
    "maxIntegrity": 60,
    "prerequisites": Object {
      "Ingot/Silicon": 15,
    },
    "size": Object {
      "X": 0.5,
      "Y": 0.5,
      "Z": 0.03,
    },
    "subtype": "BulletproofGlass",
    "time": 1,
    "type": "Component",
    "volume": 8,
  },
  Object {
    "health": 11,
    "mass": 0.2,
    "maxIntegrity": 1,
    "prerequisites": Object {
      "Ingot/Iron": 0.5,
      "Ingot/Silicon": 0.2,
    },
    "size": Object {
      "X": 0.1,
      "Y": 0.1,
      "Z": 0.02,
    },
    "subtype": "Computer",
    "time": 1,
    "type": "Component",
    "volume": 1,
  },
  Object {
    "health": 14,
    "mass": 25,
    "maxIntegrity": 20,
    "prerequisites": Object {
      "Ingot/Iron": 15,
      "Ingot/Silver": 5,
      "Ingot/Stone": 20,
    },
    "size": Object {
      "X": 0.2,
      "Y": 0.2,
      "Z": 0.2,
    },
    "subtype": "Reactor",
    "time": 8,
    "type": "Component",
    "volume": 8,
  },
  Object {
    "health": 12,
    "mass": 40,
    "maxIntegrity": 30,
    "prerequisites": Object {
      "Ingot/Cobalt": 10,
      "Ingot/Gold": 1,
      "Ingot/Iron": 30,
      "Ingot/Platinum": 0.4,
    },
    "size": Object {
      "X": 0.2,
      "Y": 0.2,
      "Z": 0.2,
    },
    "subtype": "Thrust",
    "time": 4,
    "type": "Component",
    "volume": 10,
  },
  Object {
    "health": 12,
    "mass": 800,
    "maxIntegrity": 500,
    "prerequisites": Object {
      "Ingot/Cobalt": 220,
      "Ingot/Gold": 10,
      "Ingot/Iron": 600,
      "Ingot/Silver": 5,
    },
    "size": Object {
      "X": 1,
      "Y": 0.7,
      "Z": 0.4,
    },
    "subtype": "GravityGenerator",
    "time": 10,
    "type": "Component",
    "volume": 200,
  },
  Object {
    "health": 12,
    "mass": 150,
    "maxIntegrity": 70,
    "prerequisites": Object {
      "Ingot/Iron": 60,
      "Ingot/Nickel": 70,
      "Ingot/Silver": 20,
    },
    "size": Object {
      "X": 1,
      "Y": 0.4,
      "Z": 0.4,
    },
    "subtype": "Medical",
    "time": 6,
    "type": "Component",
    "volume": 160,
  },
  Object {
    "health": 11,
    "mass": 8,
    "maxIntegrity": 15,
    "prerequisites": Object {
      "Ingot/Iron": 8,
      "Ingot/Silicon": 1,
    },
    "size": Object {
      "X": 1,
      "Y": 0.5,
      "Z": 0.4,
    },
    "subtype": "RadioCommunication",
    "time": 4,
    "type": "Component",
    "volume": 70,
  },
  Object {
    "health": 11,
    "mass": 5,
    "maxIntegrity": 4,
    "prerequisites": Object {
      "Ingot/Iron": 5,
      "Ingot/Nickel": 15,
    },
    "size": Object {
      "X": 0.15,
      "Y": 0.15,
      "Z": 0.4,
    },
    "subtype": "Detector",
    "time": 4,
    "type": "Component",
    "volume": 6,
  },
  Object {
    "health": 14,
    "mass": 15,
    "maxIntegrity": 5,
    "prerequisites": Object {
      "Ingot/Iron": 2,
      "Ingot/Silicon": 35,
    },
    "size": Object {
      "X": 0.25,
      "Y": 0.25,
      "Z": 0.03,
    },
    "subtype": "Canvas",
    "time": 4,
    "type": "Component",
    "volume": 8,
  },
  Object {
    "health": 21,
    "mass": 2,
    "maxIntegrity": 5,
    "prerequisites": Object {
      "Ingot/Magnesium": 2,
      "Ingot/Silicon": 0.5,
    },
    "size": Object {
      "X": 0.2,
      "Y": 0.2,
      "Z": 0.1,
    },
    "subtype": "Explosives",
    "time": 6,
    "type": "Component",
    "volume": 2,
  },
  Object {
    "health": 11,
    "mass": 6,
    "maxIntegrity": 1,
    "prerequisites": Object {
      "Ingot/Nickel": 3,
      "Ingot/Silicon": 6,
    },
    "size": Object {
      "X": 1,
      "Y": 0.5,
      "Z": 0.05,
    },
    "subtype": "SolarCell",
    "time": 2,
    "type": "Component",
    "volume": 12,
  },
  Object {
    "health": 14,
    "mass": 25,
    "maxIntegrity": 50,
    "prerequisites": Object {
      "Ingot/Iron": 10,
      "Ingot/Nickel": 2,
      "Ingot/Silicon": 1,
    },
    "size": Object {
      "X": 0.5,
      "Y": 0.3,
      "Z": 0.3,
    },
    "subtype": "PowerCell",
    "time": 4,
    "type": "Component",
    "volume": 40,
  },
  Object {
    "health": 14,
    "mass": 15,
    "maxIntegrity": 5,
    "prerequisites": Object {
      "Ingot/Gold": 2,
      "Ingot/Iron": 10,
    },
    "size": Object {
      "X": 0.5,
      "Y": 0.5,
      "Z": 0.03,
    },
    "subtype": "Superconductor",
    "time": 8,
    "type": "Component",
    "volume": 8,
  },
]
`;

exports[`PraisalManager with vanilla materials only should build all ingots 1`] = `10`;

exports[`PraisalManager with vanilla materials only should build all ingots 2`] = `
Array [
  "Ingot/Stone",
  "Ingot/Iron",
  "Ingot/Nickel",
  "Ingot/Cobalt",
  "Ingot/Magnesium",
  "Ingot/Silicon",
  "Ingot/Silver",
  "Ingot/Gold",
  "Ingot/Platinum",
  "Ingot/Uranium",
]
`;

exports[`PraisalManager with vanilla materials only should build all ingots 3`] = `
Array [
  Object {
    "mass": 1,
    "prerequisites": Object {
      "Ore/Stone": 1,
    },
    "subtype": "Stone",
    "time": 0.1,
    "type": "Ingot",
    "volume": 0.37,
  },
  Object {
    "mass": 1,
    "prerequisites": Object {
      "Ore/Iron": 1.4285714285714286,
    },
    "subtype": "Iron",
    "time": 0.05,
    "type": "Ingot",
    "volume": 0.127,
  },
  Object {
    "mass": 1,
    "prerequisites": Object {
      "Ore/Nickel": 2.5,
    },
    "subtype": "Nickel",
    "time": 0.66,
    "type": "Ingot",
    "volume": 0.112,
  },
  Object {
    "mass": 1,
    "prerequisites": Object {
      "Ore/Cobalt": 3.3333333333333335,
    },
    "subtype": "Cobalt",
    "time": 3,
    "type": "Ingot",
    "volume": 0.112,
  },
  Object {
    "mass": 1,
    "prerequisites": Object {
      "Ore/Magnesium": 142.85714285714286,
    },
    "subtype": "Magnesium",
    "time": 0.5,
    "type": "Ingot",
    "volume": 0.575,
  },
  Object {
    "mass": 1,
    "prerequisites": Object {
      "Ore/Silicon": 1.4285714285714286,
    },
    "subtype": "Silicon",
    "time": 0.6,
    "type": "Ingot",
    "volume": 0.429,
  },
  Object {
    "mass": 1,
    "prerequisites": Object {
      "Ore/Silver": 10,
    },
    "subtype": "Silver",
    "time": 1,
    "type": "Ingot",
    "volume": 0.095,
  },
  Object {
    "mass": 1,
    "prerequisites": Object {
      "Ore/Gold": 100,
    },
    "subtype": "Gold",
    "time": 0.4,
    "type": "Ingot",
    "volume": 0.052,
  },
  Object {
    "mass": 1,
    "prerequisites": Object {
      "Ore/Platinum": 200,
    },
    "subtype": "Platinum",
    "time": 3,
    "type": "Ingot",
    "volume": 0.047,
  },
  Object {
    "mass": 1,
    "prerequisites": Object {
      "Ore/Uranium": 100,
    },
    "subtype": "Uranium",
    "time": 4,
    "type": "Ingot",
    "volume": 0.052,
  },
]
`;

exports[`PraisalManager with vanilla materials only should build all ores 1`] = `13`;

exports[`PraisalManager with vanilla materials only should build all ores 2`] = `
Array [
  "Ore/Stone",
  "Ore/Iron",
  "Ore/Nickel",
  "Ore/Cobalt",
  "Ore/Magnesium",
  "Ore/Silicon",
  "Ore/Silver",
  "Ore/Gold",
  "Ore/Platinum",
  "Ore/Uranium",
  "Ore/Scrap",
  "Ore/Ice",
  "Ore/Organic",
]
`;

exports[`PraisalManager with vanilla materials only should build all ores 3`] = `
Array [
  Object {
    "mass": 1,
    "subtype": "Stone",
    "type": "Ore",
    "volume": 0.37,
  },
  Object {
    "mass": 1,
    "subtype": "Iron",
    "type": "Ore",
    "volume": 0.37,
  },
  Object {
    "mass": 1,
    "subtype": "Nickel",
    "type": "Ore",
    "volume": 0.37,
  },
  Object {
    "mass": 1,
    "subtype": "Cobalt",
    "type": "Ore",
    "volume": 0.37,
  },
  Object {
    "mass": 1,
    "subtype": "Magnesium",
    "type": "Ore",
    "volume": 0.37,
  },
  Object {
    "mass": 1,
    "subtype": "Silicon",
    "type": "Ore",
    "volume": 0.37,
  },
  Object {
    "mass": 1,
    "subtype": "Silver",
    "type": "Ore",
    "volume": 0.37,
  },
  Object {
    "mass": 1,
    "subtype": "Gold",
    "type": "Ore",
    "volume": 0.37,
  },
  Object {
    "mass": 1,
    "subtype": "Platinum",
    "type": "Ore",
    "volume": 0.37,
  },
  Object {
    "mass": 1,
    "subtype": "Uranium",
    "type": "Ore",
    "volume": 0.37,
  },
  Object {
    "mass": 1,
    "subtype": "Scrap",
    "type": "Ore",
    "volume": 0.254,
  },
  Object {
    "mass": 1,
    "subtype": "Ice",
    "type": "Ore",
    "volume": 0.37,
  },
  Object {
    "mass": 1,
    "subtype": "Organic",
    "type": "Ore",
    "volume": 0.37,
  },
]
`;
